new lint: `macro_metavars_in_unsafe`

y21 opened the pull request:
This implements a lint that I've been meaning to write for a while: a macro with an `expr` metavariable that is then expanded in an unsafe context. It's bad because it lets the user write unsafe code without an unsafe block.

Note: this has gone through some major rewrites, so any comment before https://github.com/rust-lang/rust-clippy/pull/12107#issuecomment-2029503545 is outdated and was based on an older version that has since been completely rewritten.


changelog: new lint: [`macro_metavars_in_unsafe`]

samueltardieu commented:
I have run your lint against the top 500 crates. Here are the results, I haven't analyzed it, but it might help you check whether it does what you intend it to.

### Reports

target/lintcheck/sources/bitvec-1.0.1/src/macros.rs:219:27 clippy::expr_metavars_in_unsafe "expanding an expression metavariable in an unsafe block"
target/lintcheck/sources/bitvec-1.0.1/src/macros.rs:244:34 clippy::expr_metavars_in_unsafe "expanding an expression metavariable in an unsafe block"
target/lintcheck/sources/bitvec-1.0.1/src/macros.rs:249:34 clippy::expr_metavars_in_unsafe "expanding an expression metavariable in an unsafe block"
target/lintcheck/sources/pyo3-0.20.2/src/types/mod.rs:185:49 clippy::expr_metavars_in_unsafe "expanding an expression metavariable in an unsafe block"
target/lintcheck/sources/static_assertions-1.1.0/src/assert_eq_size.rs:81:38 clippy::expr_metavars_in_unsafe "expanding an expression metavariable in an unsafe block"
target/lintcheck/sources/static_assertions-1.1.0/src/assert_eq_size.rs:82:62 clippy::expr_metavars_in_unsafe "expanding an expression metavariable in an unsafe block"


### Stats:

| lint                                               | count |
| -------------------------------------------------- | ----- |
| clippy::expr_metavars_in_unsafe                    |     6 |

y21 commented:
Thanks! My thoughts on the results:

## poy3
The pyo3 case looks like a real case to me. `pyo3::pyobject_native_static_type_object!((*std::ptr::null_mut::<u8>()))(());
` invokes undefined behavior with no unsafe block in sight.

## bitvec
Looks like a false positive. This is basically:
```rs
macro_rules! x {
  ($v:expr) => {
    $v; unsafe { $v; };
  }
}
```
It's technically expanding `$v` in an unsafe block, but it's still not possible to write unsafe without an unsafe block at callsite because it's expanded twice. The first use of `$v` is outside of an unsafe block.
I think this is fixable.

## static_assertions
The two static_assertions cases I already saw and mentioned in my description. I suppose those are technically FPs because they're in a closure that is never called, thus never reached, so it doesn't matter what kinds of unsafety is done, but I don't think this is something we can detect. I want to say that if this is an uncommon case, it could be something that can be `#[allow]`ed, and they already do have multiple `#[allow]`s above it.

y21 commented:
> `pyo3::pyobject_native_static_type_object!((*std::ptr::null_mut::<u8>()))(());`

Oh sorry this specifically is not UB (anymore?). pyo3's macro is using `addr_of_mut!`, which creates a place. I'm pretty sure that the documentation here was updated, it used to say that even dereferencing a null pointer in `addr_of_mut!` is undefined behavior, but not anymore.
Anyways, there are still [other ways](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=260d671c54fe475d6ec6ab3c5b736785) to create UB with this macro

But I also completely missed that the macro is annotated with `#[doc(hidden)]`, so this is probably also a FP then, since they're not considered to be part of the public API

y21 commented:
Fixed those 2 FPs. The more logic I add to this, though, I feel like this isn't a good idea, since this is effectively implementing an unsafety checker at the lexer level...
I think I'm going to experiment with different ways to do this.
Maybe it's possible to use the `rustc_expand` crate.
Another way I can think of is, instead of looking at the macro definition, look at the expansions of local crates. The upside is that we have an actual AST/HIR to work with.

y21 commented:
Finally got around to rewriting this by having it look at local macro expansions, rather than trying to parse the macro definition tokenstream.
It did make the code significantly simpler, but has a few downsides:

- ideally we would tell the user the metavariable that was expanded in an unsafe block, but there is (afaik) no way to get the span of the `$ident` in the macro definition, given its expansion. I tried many hacks (such as getting the parent of the `$ident` expansion, match on the node kind and retokenize everything in between to find something that looks like `$ident` but there are cases where it doesn't work)
  - this currently works around that limitation by mentioning the unsafe block instead
- as a consequence, we have no way of checking that the metavariable is *actually* an `:expr` metavariable
  - making this work probably requires parsing *some* of the macro tokenstream? unless there's something in the rustc crates
- this won't catch macros that are never invoked locally, which can be seen by lintcheck (I'm getting 0 hits now). In practice however I think the fact that crates usually have tests for macros in the same crate makes this less bad.
- since a macro can expand the same metavariable practically anywhere (even across multiple functions), we need to keep track of spans so that we can undo warnings
  - we really only need spans from macros, so this cuts the number of spans that need to be stored quite a bit

... so for now I put this in the nursery category. I'm not sure which one of these can be worked around or improved. Will think about it a bit more.

bors commented:
:umbrella: The latest upstream changes (presumably #12160) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

xFrednet commented:
@GuillaumeGomez would you mind taking a look at this PR?

It has some comments, which you probably first need to catch up on.

r? xFrednet

GuillaumeGomez reviewed (approved):
Code is good and well documented. Nice work! Looks ready to go for me.

y21 commented:
Yeah the PR history is a bit of a mess, it went through a few rewrites :D Though half the comments are no longer really relevant.

It started by inspecting/"parsing" the `TokenStream` of a macro definition to find unsafe blocks and `$expr`s (since it has not AST) which very quickly got out of hand when trying to handle edge cases, so instead of doing that we simply look for expanded expressions in the HIR.

I think the main "blocker" more or less is that this currently does not check that the expanded metavariable is an expression, contrary to the lint name:
```rs
macro_rules! m {
  ($v:literal) => { unsafe { $v } }
}
m!(1)
```
This is linted even though `$v:literal`.

But... maybe it's fine and we could just rename the lint to not mention "expr_metavars"? What about something like `macro_metavars_in_unsafe`?

xFrednet commented:
> But... maybe it's fine and we could just rename the lint to not mention "expr_metavars"? What about something like `macro_metavars_in_unsafe`?

This sounds good to me. Then it'll also catch statements and other tokens :)

xFrednet commented on clippy_lints/src/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,241 @@
+    /// Instead, this lint looks at expansions of macros defined in the same crate.
+    /// This leads to false negatives when a macro is never actually invoked.
+    ///
+    /// ### Example
```
Since the lint explicitly checks for cases, where the metavariable is also used in safe and unsafe blocks, I wonder if that could be a valid suggestion. It shouldn't be the default, but if someone really needs the unsafe block at this location for some reason, they could just add a `if false {$expr}` somewhere.

I'm considering if we might want to mention this in the lint description, but it feels like that's just a hack and not a nice solution. If you agree, you can just resolve this :D

y21 replied:
I'm not a big fan of that hack but yeah I suppose it is worth mentioning in the description anyway as a last resort if there is no other way for people to restructure their macros. Added to the documentation.

xFrednet commented on clippy_lints/src/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,241 @@
+
+        for (id, span, def_id) in bad_unsafe_blocks {
+            if let Some(def_id) = def_id
+                && (cx.effective_visibilities.is_exported(def_id) || cx.tcx.has_attr(def_id, sym::macro_export))
+                && !cx.tcx.is_doc_hidden(def_id)
```
This also seems like something which should be configurable. I would like it to lint also on internal macros. On the other hand, I understand that this might result in more lint triggers.

If we keep this behavior, could you add it to the lint description?

y21 replied:
Added that optional configuration in the last commit.

xFrednet commented on clippy_lints/src/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,241 @@
+                    EXPR_METAVARS_IN_UNSAFE,
+                    id,
+                    span,
+                    "this unsafe block in a macro expands `expr` metavariables",
```
```suggestion
                    "this unsafe block in a macro expands metavariables",
```

xFrednet commented on tests/ui/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,228 @@
+#[macro_export]
+macro_rules! allow_works {
+    ($v:expr) => {
+        #[allow(clippy::expr_metavars_in_unsafe)]
```
Could you also add a test for `#[expect]`? That will require the `lint_reasons` feature.

xFrednet commented on tests/ui/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,228 @@
+}
+
+fn main() {
+    allow_works!(1);
```
Could you also add tests, where the macro is safe, but the user passes `unsafe { 1 }` in as a meta variable?

xFrednet reviewed (commented):
Very nice implementation! The structure and docs make the code very understandable. I have several nit-picky comments, but hopefully nothing that will require a lot of work. :D

y21 commented on clippy_lints/src/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,241 @@
+    /// Looks for macros that expand metavariables in an unsafe block.
+    ///
+    /// ### Why is this bad?
+    /// This is unsound: it allows the user of the macro to write unsafe code outside of an
+    /// unsafe block at callsite, potentially invoking undefined behavior in safe code.
```
> That in itself is not unsound

Hmm, I was going to refer back to [UCG](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library), which *to me* made this sound like it would clearly be unsound, practically speaking: the user doesn't have unsafe blocks in their code and the macro is in the public API of a library, but I guess it depends on whether you consider code inside of a macro-expanded unsafe block still be safe code.

According to https://github.com/rust-lang/unsafe-code-guidelines/issues/278 , it looks like it's not entirely clear if this is sound or not, so I guess it's better to just avoid mentioning soundness, then.

y21 replied:
I reworded this a bit in the last commit to avoid mentioning the terms (un)soundness. Lmk what you think

y21 commented:
Thanks for reviewing, I'll address the comments eventually, but I will be busy for a couple more weeks, so it might take a while.

y21 commented on clippy_lints/src/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,241 @@
+        if let ExprKind::Block(block, _) = e.kind
+            && let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules
+            && !ctxt.is_root()
+            && ctxt.outer_expn_data().macro_def_id.is_some_and(DefId::is_local)
```
Yea, that sounds like a good idea, however I'd like to do that separately also. It seems like that would require a fair amount of changes, but should be doable.

y21 commented on clippy_lints/src/expr_metavars_in_unsafe.rs:
```diff
@@ -0,0 +1,241 @@
+        let from_expn = s.span.from_expansion();
+        if from_expn {
+            self.expn_depth += 1;
+        }
```
Can you elaborate why this is weird or what you would rename this field to to make it less misleading? Or what exact test case you'd want?

To clarify, It's okay for this to falsely return true and it doesn't matter for correctness if we increment the counter multiple times for the same macro or for unrelated macros, *as long as* we arrive back at 0 when we leave all expansion spans in that function and that the counter is `>0` if there are any enclosing macro spans.

xFrednet replied:
> To clarify, It's okay for this to falsely return true and it doesn't matter for correctness if we increment the counter multiple times for the same macro or for unrelated macros, _as long as_ we arrive back at 0 when we leave all expansion spans

That's a good point. I can't think of an example that would violate the implementation. Then let's leave it at this :)

y21 commented on clippy_config/src/conf.rs:
```diff
@@ -609,6 +609,10 @@ define_Conf! {
+    /// Lint: MACRO_METAVARS_IN_UNSAFE.
+    ///
+    /// Whether to also emit warnings for unsafe blocks with metavariable expansions in **private** macros.
+    (warn_unsafe_macro_metavars_in_private_macros: bool = false),
```
This is quite lengthy, if someone has a better and more concise name for this, do tell

xFrednet replied:
I like the name, it's lengthy but descriptive :+1:

xFrednet commented on tests/ui-toml/macro_metavars_in_unsafe/default/test.stderr:
```diff
@@ -0,0 +1,187 @@
+error: this unsafe block in a macro expands metavariables
```
The message doesn't read fluently to me. What do you think about:
```suggestion
error: this unsafe block expands metavariables
```
Or maybe better
```suggestion
error: this macro expands metavariables in an unsafe block
```

y21 replied:
:+1: I agree, the second suggestion sounds good

xFrednet reviewed (approved):
Tiny error message NIT, but the rest looks good to me.

You can r=me after the changed message, if you take one of my suggestions. :D

y21 commented:
I created an FCP thread on zulip for this. Will squash when that's complete.

xFrednet commented:
Right, I forgot about the FCP. Thank you for remembering :heart:

blyxyas commented on clippy_lints/src/macro_metavars_in_unsafe.rs:
````diff
@@ -0,0 +1,256 @@
+    /// // This won't compile:
+    /// assert_eq!(*first!(std::hint::unreachable_unchecked() as &[i32]), 1);
+    /// ```
+    #[clippy::version = "1.77.0"]
````
```suggestion
    #[clippy::version = "1.80.0"]
```

blyxyas replied:
I'm thinking of adding a test for this kind of case, something like the "CURRENT_RUSTC_VERSION" weird thing that exists in the compiler but without having a weird variable.

When I said "I'm thinking of" I meant "I'm working on it right now"

y21 replied:
Maybe this'd be worth bringing up on zulip. At one point someone told me that it's not problematic if this is wrong since it gets corrected in the changelog PRs. But there is the issue that in the short timeframe while the lint is still in nightly/beta, in the documentation it can "look like" the lint already exists on stable because the version hasn't been corrected yet (this happened in #12559 for example).

But anyway I will change this version attribute to the latest one once fcp is complete in case this gets delayed to 1.81 :+1:

blyxyas replied:
We had the same idea of talking about it on Zulip :sweat_smile:
https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Using.20CURRENT_RUSTC_VERSION.20in.20clippy.3A.3Aversion.3F/near/437353709

xFrednet commented:
Since no comments have come up during the FCP it should be safe to merge this now.

@y21 can you resolve the last comment regarding the version? Then you can `r=me` =^.^=

bors commented:
:umbrella: The latest upstream changes (presumably #12620) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

y21 commented:
Changed the version atribute to 1.80 and rebased to fix conflicts + squashed. (had to force push twice since the first commit still had the old lint name in its commit message which was slightly misleading)

@bors r=xFrednet

bors commented:
:pushpin: Commit 9747c80644fc9f705029c254ca1d14668a72f904 has been approved by `xFrednet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=xFrednet 9747c80644fc9f705029c254ca1d14668a72f904 -->
<!-- homu: {"type":"Approved","sha":"9747c80644fc9f705029c254ca1d14668a72f904","approver":"xFrednet","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 9747c80644fc9f705029c254ca1d14668a72f904 with merge 50e1065aed5e90300433af654acdb3334962bdc9...
<!-- homu: {"type":"BuildStarted","head_sha":"9747c80644fc9f705029c254ca1d14668a72f904","merge_sha":"50e1065aed5e90300433af654acdb3334962bdc9"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052382717/job/24870001396), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052382721/job/24870004032), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052382720/job/24870089525)
Approved by: xFrednet
Pushing 50e1065aed5e90300433af654acdb3334962bdc9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"xFrednet","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052382720/job/24870089525","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052382717/job/24870001396","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052382721/job/24870004032"},"merge_sha":"50e1065aed5e90300433af654acdb3334962bdc9"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 1 review requesting changes and 1 approving review by reviewers with write access.

xFrednet commented:
The merge conflict had perfect timing xD

@bors retry

bors commented:
:hourglass: Testing commit 9747c80644fc9f705029c254ca1d14668a72f904 with merge 323eca63579bef4a09ad9c990256a7fab588e061...
<!-- homu: {"type":"BuildStarted","head_sha":"9747c80644fc9f705029c254ca1d14668a72f904","merge_sha":"323eca63579bef4a09ad9c990256a7fab588e061"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052614264/job/24870510842), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052614262/job/24870511942), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052614266/job/24870585409)
Approved by: xFrednet
Pushing 323eca63579bef4a09ad9c990256a7fab588e061 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"xFrednet","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052614266/job/24870585409","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052614264/job/24870510842","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052614262/job/24870511942"},"merge_sha":"323eca63579bef4a09ad9c990256a7fab588e061"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 1 review requesting changes and 1 approving review by reviewers with write access.

xFrednet commented:
@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors commented:
:pushpin: Commit 9747c80644fc9f705029c254ca1d14668a72f904 has been approved by `xFrednet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=xFrednet 9747c80644fc9f705029c254ca1d14668a72f904 -->
<!-- homu: {"type":"Approved","sha":"9747c80644fc9f705029c254ca1d14668a72f904","approver":"xFrednet","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 9747c80644fc9f705029c254ca1d14668a72f904 with merge a4a1a7365d60280b83d83374fda97742fb959239...
<!-- homu: {"type":"BuildStarted","head_sha":"9747c80644fc9f705029c254ca1d14668a72f904","merge_sha":"a4a1a7365d60280b83d83374fda97742fb959239"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052780684/job/24870856618), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052780686/job/24870859948), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/9052780681/job/24870928805)
Approved by: xFrednet
Pushing a4a1a7365d60280b83d83374fda97742fb959239 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"xFrednet","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052780681/job/24870928805","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052780684/job/24870856618","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9052780686/job/24870859948"},"merge_sha":"a4a1a7365d60280b83d83374fda97742fb959239"} -->
